src/main/java/com/opendxl/databus/consumer/DatabusPushConsumer.java

```diff
@@ -0,0 +1,442 @@
+     * or stop reading messages. Messages read from Databus are sent to listener in an separated and parallel thread.
+     * While the listener is working on messages received from Databus, the Push Consumer pauses topic-partitions
+     * assigned to it and sends heartbeats the Kafka broker in background.
+     * Once the listener has finished, it should returns a
+     * {@link DatabusPushConsumerListenerResponse} value or an Exception in case something was wrong.
+     * Then the main loop will act accordingly. The main loop keep a {@code databusPushConsumerFuture} future argument
+     *
+     * @param databusPushConsumerFuture a future instance shared which is updated to let the
+     *                                  SDK Databus client know about the process status.
+     * @param timeout                   The time, in milliseconds, spent waiting in poll if data is not available.
+     *                                  If 0, returns
+     *                                  immediately with any records that are available now. Must not be negative.
+     */
+    private void push(final DatabusPushConsumerFuture databusPushConsumerFuture,
+                      final Duration timeout) {
+
+        LOG.info("Consumer " + super.getClientId() + " start");
+        try {
+
+            // Main loop. It keeps the reading-pushing mechanism running until either a close() was called or
+            // the listener asks to stop or the listener throws an exception.
+            while (!stopRequested.get()) {
+
+                // It stores the current consumer position based on the partitions assigned.
+                // This is the offset position per topicPartition before polling messages from Databus.  Position
+                // will be used to rewind the consumer in case the listener returns RETRY enum value.
+                final Set<TopicPartition> assignment = super.assignment();
+                final Map<TopicPartition, Long> lastKnownPositionPerTopicPartition =
+                        getCurrentConsumerPosition(assignment);
+
+                // poll records from Databus
+                ConsumerRecords records = super.poll(timeout);
+                LOG.info("Consumer " + super.getClientId() + " number of records read: " + records.count());
+
+                if (records.count() > 0) {
+                    super.pause(assignment);
+                    LOG.info("Consumer " + super.getClientId() + " is paused");
+                } else {
+                    continue;
+                }
+
+                // It calls the listener in a separated thread.
+                listenerFuture = runListenerAsync(databusPushConsumerFuture, records);
+
+                DatabusPushConsumerListenerResponse onConsumeResponse ;
+                boolean listenerIsFinished = false;
+
+                // This loop wait for listener the resutl or manages exceptions that listener might throws.
```

Comment:
Cosmetic typo:  This loop wait_s_ for the listener result and it also manages exceptions that listener might throw.